external/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisDStream.scala

```diff
@@ -0,0 +1,559 @@
+     * credential chain is used to authenticate to STS and no external ID is used.
+     *
+     * @param stsRoleArn ARN of the IAM role to assume via STS
+     * @param stsSessionName Name to identify STS sessions by
+     * @return Reference to this [[KinesisDStream.Builder]]
+     */
+    def kinesisSTSCredentials(stsRoleArn: String, stsSessionName: String): Builder[T] = {
+      kinesisCredsProvider = buildSTSCredentials(stsRoleArn, stsSessionName)
+      this
+    }
+
+    /**
+     * Assume an STS role in order to authenticate to the Kinesis service. The provided
+     * AWS credential keypair is used to authenticate to STS and no external ID is used.
+     *
+     * @param stsRoleArn ARN of the IAM role to assume via STS
+     * @param stsSessionName Name to identify STS sessions by
+     * @param awsAccessKeyId AWS access key ID for authenicating to STS
+     * @param awsSecretKey AWS secret key for authenticating to STS
+     * @return Reference to this [[KinesisDStream.Builder]]
+     */
+    def kinesisSTSCredentials(
+        stsRoleArn: String,
+        stsSessionName: String,
+        awsAccessKeyId: String,
+        awsSecretKey: String): Builder[T] = {
+      kinesisCredsProvider = buildSTSCredentials(
+        stsRoleArn,
+        stsSessionName,
+        awsKeypair = Option(awsAccessKeyId, awsSecretKey))
+      this
+    }
+
+    /**
+     * Assume an STS role in order to authenticate to the Kinesis service. The default
+     * credential chain is used to authenticate to STS and the provided external ID is
+     * validated against the value in the assumed role's policy (if one is specified).
+     *
+     * @param stsRoleArn ARN of the IAM role to assume via STS
+     * @param stsSessionName Name to identify STS sessions by
+     * @param stsExternaId External ID used by STS to validate IAM role trust policy
+     * @return Reference to this [[KinesisDStream.Builder]]
+     */
+    def kinesisSTSCredentials(
+        stsRoleArn: String,
+        stsSessionName: String,
+        stsExternalId: String): Builder[T] = {
+      kinesisCredsProvider = buildSTSCredentials(
+        stsRoleArn,
+        stsSessionName,
+        stsExternalId = Option(stsExternalId))
+      this
+    }
+
+    /**
+     * Assume an STS role in order to authenticate to the Kinesis service. The provided
+     * AWS credential keypair is used to authenticate to STS and the provided external ID
+     * is validated against the value in the assumed role's policy (if one is specified).
+     *
+     * @param stsRoleArn ARN of the IAM role to assume via STS
+     * @param stsSessionName Name to identify STS sessions by
+     * @param awsAccessKeyId AWS access key ID for authenicating to STS
+     * @param stsExternaId External ID used by STS to validate IAM role trust policy
+     * @param awsSecretKey AWS secret key for authenticating to STS
+     * @return Reference to this [[KinesisDStream.Builder]]
+     */
+    def kinesisSTSCredentials(
```

Comment:
With this many methods for the credential provider, I feel we need a credential provider builder. I wouldn't want to re-enter everything again between `dynamoDb` and `cloudWatch`. If I want to keep them separate from the `kinesis` credentials